File: src/main/scala/arimitsu/sf/akka/cqlclient/message/Prepare.scala
package arimitsu.sf.akka.cqlclient.message

import arimitsu.sf.cql.v3.messages.Result.Kind
import scala.concurrent.Promise
import arimitsu.sf.cql.v3.Frame
import arimitsu.sf.cql.v3.messages.{Prepared, Result}
import java.nio.ByteBuffer

/**
 * Created by sxend on 2014/06/11.
 */
case class Prepare(query: String, promise: Promise[Prepared]) extends Message(promise) {

  override def process(frame: Frame): Unit = {
    val result = Result.Factory.fromBuffer(ByteBuffer.wrap(frame.body))
    result.getKind match {
      case Kind.PREPARED => promise.success(result.asInstanceOf[Prepared])
      case _ => promise.failure(new RuntimeException("invalid prepare response"))
    }
  }


}

File: src/main/java/arimitsu/sf/cql/v3/messages/Prepare.java
package arimitsu.sf.cql.v3.messages;

import arimitsu.sf.cql.v3.Flags;
import arimitsu.sf.cql.v3.Frame;
import arimitsu.sf.cql.v3.Header;
import arimitsu.sf.cql.v3.Opcode;
import arimitsu.sf.cql.v3.Version;
import arimitsu.sf.cql.v3.util.Notation;

/**
 * Created by sxend on 14/06/07.
 */
public class Prepare implements Request {
    public final short streamId;
    public final Flags flags;
    public final String query;

    public Prepare(short streamId, Flags flags, String query) {
        this.streamId = streamId;
        this.flags = flags;
        this.query = query;
    }

    @Override
    public Frame toFrame() {
        return new Frame(new Header(Version.REQUEST, flags, streamId, Opcode.PREPARE), Notation.toLongString(query));
    }
}
